app/models/metric/ci_mixin/capture.rb

```diff
@@ -80,10 +80,12 @@ def perf_capture_queue(interval_name, options = {})
       :state       => ['ready', 'dequeue'],
     }
 
+    messages = MiqQueue.where(queue_item).index_by { |m| [m.method_name, m.args] }
     items.each do |item_interval, *start_and_end_time|
       # Should both interval name and args (dates) be part of uniqueness query?
       queue_item_options = queue_item.merge(:method_name => "perf_capture_#{item_interval}")
       queue_item_options[:args] = start_and_end_time if start_and_end_time.present?
+      next if messages[queue_item_options.values_at(:method_name, :args)]
       MiqQueue.put_or_update(queue_item_options) do |msg, qi|
```

Comment:
> When you put an item into the queue, should you also be updating your in-memory index?

In theory yes. Not needed in this case.

> My only other concern is that put_or_update also considers default values for certain columns (e.g. priority), which you are ignoring in your .where call. I'm not sure they are important in this particular case though.

The default values for certain columns (like priority) are not used in where() clause of put_or_update. Only defaults for `queue_name, state, zone` are used in the `where()` clause within put_or_update. That means we are fine.

>I wonder if it's better to use MiqQueue.peek instead?

I cannot see how we could leverage `peek` for this.


Anyway, the only thing that remains to be decided is the failed specs. It uncovers greater problem (even more room for perf improvement).